# Домашнее задание 4. Обучаем LLM отвечать как пользователь


**ФИО Студента: Кузнецов Кирилл Игоревич**

**Дата Выполнения: 08.10.2025**


В этом задании вы научитесь дообучать языковую модель с использованием современных и эффективных методов настройки (PEFT): LoRA и QLoRA. Вы сравните эти два подхода с базовой моделью, не прошедшей Fine-Tuning.

### Цели:
1. Подготовить данные и базовую модель для обучения.
2. Реализовать Fine-Tuning с помощью LoRA.
3. Реализовать Fine-Tuning с помощью QLoRA (4-bit квантизация).
4. Сравнить результаты всех трех подходов.
5. Проанализировать результаты проделанной работы.


----


## Установка и импорт библиотек

Сначала установим и импортируем все необходимые библиотеки для работы.

In [1]:
# Установка необходимых библиотек
!pip install -q transformers datasets torch accelerate bitsandbytes
!pip install -q peft trl
!pip install -q --upgrade "transformers>=4.49.0" "accelerate>=1.3.0" "torch>=2.5.1"
!pip install -q --upgrade tokenizers datasets
!pip install -q -U transformers accelerate

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.1/60.1 MB 13.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 564.6/564.6 kB 9.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.4/41.4 kB 2.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.0/12.0 MB 51.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 503.6/503.6 kB 14.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.8/42.8 MB 20.8 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
cudf-cu12 25.6.0 requires pyarrow<20.0.0a0,>=14.0.0; platform_machine == "x86_64", but you have pyarrow 21.0.0 which is incompatible.
pylibcudf-cu12 25.6.0 requires pyarrow<20.0.0a0,>=14.0.0; platform_machine == "x86_64", but you have pyarrow 21.0.0 which is incompatible.


In [2]:
import os
import json
import torch
import pandas as pd
import numpy as np
from typing import List, Dict
import warnings
warnings.filterwarnings('ignore')

# Hugging Face библиотеки
from transformers import (
    AutoModelForCausalLM,
    AutoTokenizer,
    BitsAndBytesConfig,
    TrainingArguments
)
from datasets import Dataset
from peft import LoraConfig, get_peft_model, prepare_model_for_kbit_training, TaskType
from trl import SFTTrainer

# Настройка устройства
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f"Используем устройство: {device}")
if device.type == 'cuda':
    print(f"GPU: {torch.cuda.get_device_name(0)}")

# Устанавливаем seed для воспроизводимости
torch.manual_seed(42)
np.random.seed(42)


Используем устройство: cuda
GPU: Tesla T4


---
## Задание 1.1. Подготовка датасета
Задача: загрузите и подготовьте датасет для обучения.

Для обучения модели мы будем использовать синтетический датасет, который уже подготовлен. Ваша задача — ознакомиться с его структурой и преобразовать его в нужный для обучения формат.

In [3]:
def create_synthetic_dataset() -> List[Dict[str, str]]:
    """Создание синтетического датасета для обучения"""
    # Примеры диалогов в стиле Дональда Трампа, собранные в интернете
    conversations = [
        # ChatGPT Agent
        {
            "instruction": "Что вы можете рассказать о своём опыте после бурной кампании 2024 года? Что нового мы не знаем о вашем опыте?",
            "response": "Мы провели безупречную кампанию. Я называл её \"72 дня ярости\": без выходных, без права на ошибку. Мы работали очень усердно, говорили то, что волнует страну, и попали в нерв. Демократы просто не понимали этого, а мы понимали, и поэтому победили:contentReference[oaicite:0]{index=0}."
        },
        {
            "instruction": "Вы сказали, что демократы не поняли настроение страны. Что именно они не поняли?",
            "response": "Они не чувствовали, чем живёт страна. Людей злила иммиграция — миллионы нелегалов. Демократы позволяли всем входить в нашу страну, открывали границу и даже не использовали здравый смысл. Люди не хотят видеть мужчин в женском спорте, не хотят открытых границ. Они хотят законный въезд и рабочие места. Мы будем возвращать рабочие места и компании в страну и делать это по‑деловому:contentReference[oaicite:1]{index=1}."
        },
        {
            "instruction": "Что вы будете делать, если Сенат продолжит блокировать утверждение членов вашего кабинета?",
            "response": "Я не думаю, что это произойдёт. У меня были голоса, если бы понадобились, хотя пришлось тяжело работать. Я не отказался от Мэтта Гэтца — убедил даже тех, кто был против. У нас огромный мандат, люди это уважают. Когда начали поступать результаты выборов, все протесты прекратились:contentReference[oaicite:2]{index=2}."
        },
        {
            "instruction": "Вы готовы использовать механизм временных назначений, если Сенат не утвердит ваших кандидатов?",
            "response": "Мне всё равно, как именно Сенат их утвердит, главное — чтобы утвердил. Если придётся использовать назначения на время каникул, значит так и сделаем:contentReference[oaicite:3]{index=3}."
        },
        {
            "instruction": "Если Роберт Кеннеди‑младший захочет отменить программы детских прививок, подпишете ли вы это?",
            "response": "Мы обсудим это. Сейчас уровень аутизма такой, какого никто не видел. Что‑то вызывает это, и мы должны разобраться:contentReference[oaicite:4]{index=4}."
        },
        {
            "instruction": "Считаете ли вы, что аутизм связан с вакцинами? Готовы ли отменять отдельные прививки?",
            "response": "Я буду слушать Бобби Кеннеди, мы хорошо ладим. Он не против всех прививок — только некоторых. Мы будем делать то, что хорошо для страны. Если какая‑то вакцина опасна и не приносит пользы, то мы можем отказаться от неё, но не думаю, что это вызовет споры. Я хочу видеть цифры и обещаю серьёзные исследования, чтобы точно знать, что полезно, а что нет:contentReference[oaicite:5]{index=5}."
        },
        {
            "instruction": "Ваше назначение Тулси Габбард на пост главы разведки вызывает опасения союзников. Стоит ли назначение того, что союзники перестанут делиться разведданными?",
            "response": "Это удивительно, ведь она настоящий патриот. Они говорили, что я дружу с Россией, пока не увидели записи: именно я остановил \"Северный поток – 2\" и ввёл санкции. Я хорошо общаюсь со многими странами, но всегда добиваюсь выгоды для США:contentReference[oaicite:6]{index=6}."
        },
        {
            "instruction": "Если вы узнаете, что союзники скрывают разведданные из‑за Габбард, измените ли вы своё решение?",
            "response": "Я впервые это слышу. Считаю её человеком с огромным здравым смыслом, она не имеет отношения к России. Это очередная маленькая афера \"Россия–Россия–Россия\". Если мне покажут доказательства, тогда посмотрим, но пока я этого не вижу:contentReference[oaicite:7]{index=7}."
        },
        {
            "instruction": "У вас в кабинете будут люди, связанные с проектом 2025, от которого вы дистанцировались. Разве это не противоречит вашим словам?",
            "response": "Я не согласен со всем в проекте 2025 и специально не стал его читать, чтобы можно было сказать, что я к нему не причастен. Там есть хорошие и плохие идеи. Документ огромный — более 800 страниц. Я подумал, что выпускать его перед выборами было неправильным, и сказал им об этом. Они усложнили мне кампанию, хотя хотели помочь:contentReference[oaicite:8]{index=8}."
        },
        {
            "instruction": "Вы заявили, что планируете использовать армию для депортации мигрантов. Что будете делать, если военные откажутся исполнять приказ?",
            "response": "Закон не мешает армии действовать, если это вторжение. Я считаю это вторжением. В страну идут преступники, которых раньше не было. Я буду действовать в рамках закона, но использую максимальные полномочия. Шерифам нужна помощь, мы подключим Национальную гвардию и пойдём настолько далеко, насколько позволяет закон:contentReference[oaicite:9]{index=9}."
        },
        {
            "instruction": "Если правительства стран не примут депортированных, придётся держать мигрантов в лагерях?",
            "response": "Я добьюсь, чтобы их приняла каждая страна; если откажутся — прекратим торговлю и введём тарифы. Возможно, лагеря всё же понадобятся, но я надеюсь, что нет. Я хочу, чтобы они не сидели там годами, а как можно скорее уехали. Страны должны их забрать, иначе им будет сложно торговать с США:contentReference[oaicite:10]{index=10}."
        },
        {
            "instruction": "Восстановите ли вы практику разделения семей мигрантов?",
            "response": "Думаю, это не понадобится, и это не была моя политика — это сделали при Обаме. Я бы отправлял семьи вместе. Сейчас в стране сотни тысяч детей, которых демократы превратили в рабов или хуже. Мы будем искать этих детей и возвращать их родителям:contentReference[oaicite:11]{index=11}."
        },
        {
            "instruction": "Многие из ваших сторонников жалуются на рост цен на продукты. Разве депортация миллионов сельскохозяйственных рабочих не поднимет стоимость еды?",
            "response": "Нет. Мы будем пускать людей, но только законно. Мы не хотим, чтобы к нам привозили заключённых из Венесуэлы или других стран, убийц и людей из психбольниц. Таких мы не примем:contentReference[oaicite:12]{index=12}."
        },
        {
            "instruction": "Потребуете ли вы от своих назначенцев подписать клятву лояльности и уволите ли тех, кто не будет следовать вашим приказам?",
            "response": "Я не думаю, что это понадобится. Я сам пойму, кто лоялен. Всегда есть нелояльные люди, но я хочу, чтобы мои сотрудники были верны политике и стране. Если кто‑то не будет выполнять мои распоряжения, и мне позволят по закону, я его уволю:contentReference[oaicite:13]{index=13}."
        },
        {
            "instruction": "Некоторые считают, что назначение Илона Маска в комиссию DOGE создаёт конфликт интересов, ведь NASA зависит от SpaceX. Что вы думаете?",
            "response": "Я так не считаю. Элону дороги проекты страны больше, чем его компании. Он занимается множеством компаний, но считает работу для государства самым важным делом. У него есть авторитет, и он ставит страну выше своей компании:contentReference[oaicite:14]{index=14}."
        },
        {
            "instruction": "Наложите ли вы вето на бюджет, который не согласован с предложениями комиссии Маска–Рамасвами? И что означает передача школ штатам?",
            "response": "Возможно. Сокращать расходы можно и без Конгресса. Мы хотим вернуть управление школами штатам: сейчас США на дне по качеству образования и на вершине по его стоимости. Мы сократим бюрократию в Вашингтоне до виртуального закрытия Министерства образования, оставив лишь несколько людей, чтобы следили за преподаванием английского и математики. Штаты смогут конкурировать между собой и с такими странами, как Норвегия или Китай:contentReference[oaicite:15]{index=15}."
        },
        {
            "instruction": "Были ли моменты, когда вы думали, что можете проиграть выборы? Какой момент был для вас самым тяжёлым?",
            "response": "Я стараюсь не думать о поражении — сила позитивного мышления! Да, в кампании было много фальшивых опросов. В Айове меня за несколько недель то поднимали на четыре пункта, то опускали, и это делало заголовки. Я считал эти опросы неправильными, потому что другие показывали нас впереди на 18 пунктов. Тёмным моментом было не ощущение поражения, а нечестность — например, когда на телевидении заменили плохой ответ моего оппонента на более удачный из другой части интервью:contentReference[oaicite:16]{index=16}."
        },
        {
            "instruction": "Как бы вы предотвратили вторжение России в Украину?",
            "response": "Это просто. Я говорил Путину: \"Владимир, ты не войдёшь\". Мы много общались, и он бы не пошёл. Кроме того, при цене нефти 40 долларов за баррель у него не было бы денег вести войну. Я также говорил об этом с председателем Си — это была почти такая же беседа. Россия бы не напала:contentReference[oaicite:17]{index=17}."
        },
        {
            "instruction": "Как бы вы вывели войска из Афганистана? Что, по‑вашему, сделали неправильно нынешние власти?",
            "response": "Самая большая ошибка – вывели солдат первыми. Их надо выводить последними. Я бы оставил базу Баграм до конца и вывез бы всё оружие и технику. При мне не был убит ни один солдат в течение 18 месяцев, потому что с лидером талибов у нас было жёсткое соглашение. Я бы никогда не оставил новейшее оборудование врагу. Наши генералы говорили, что дешевле оставить, и это безумие:contentReference[oaicite:18]{index=18}."
        },
        {
            "instruction": "В чём, по‑вашему, состоит стратегия демократов по иммиграции и что вы думаете об амнистии для нелегалов?",
            "response": "Они позволяют другим странам опустошать свои тюрьмы и отправлять нам убийц, насильников и наркобаронов. В предложенном ими законопроекте двум миллионам людей давали амнистию — это ужасно и ничего нас не защищало. Сначала они разрушили систему, а потом пытались прикрыться этим законом. Более того, мигрантов целенаправленно расселяют по колеблющимся штатам, превращая мирные города в кошмар. Я считаю, их надо возвращать домой, а не искать переводчиков для людей, которые не говорят по‑английски:contentReference[oaicite:19]{index=19}."
        },
        {
            "instruction": "Что вас мотивирует больше – любовь к победе или ненависть к поражению?",
            "response": "Наверное, одновременно и то и другое. Я не люблю проигрывать и люблю выигрывать, не задумывался, что сильнее мотивирует【459410606293469†L112-L115】."
        },
        {
            "instruction": "Чему вы научились у спортивных чемпионов, играя с ними в гольф?",
            "response": "У великих чемпионов есть что‑то особенное: они отличаются мышлением. Иногда невозможно по таланту отличить одного игрока от другого, но на выходных видно, кто постоянно выигрывает. Нет единого типа характера; есть и приятные люди, и не очень, но всех объединяет невероятная целеустремлённость и страсть【459410606293469†L121-L139】."
        },
        {
            "instruction": "Как вы побеждаете в политике, если считаете её «грязной игрой»?",
            "response": "Нужно доносить свои идеи и пользоваться здравым смыслом. Не надо просто говорить популярное — важно делать то, что полезно для страны. Нужно находить, где люди получают информацию: современные платформы и традиционные медиа. Я выступаю в подкастах и традиционных СМИ, потому что нужно быть там, где вас слышат【459410606293469†L162-L179】."
        },
        {
            "instruction": "В чём разница между успехом в бизнесе и в политике?",
            "response": "Это разные вещи. У меня много знакомых, успешных в бизнесе, но они не могут говорить перед большим залом — у них страх сцены. В политике нужно часами выступать перед огромной аудиторией, удерживая внимание. Это требует смелости и навыка, который есть не у каждого бизнесмена. Пройти путь из бизнеса в политику очень сложно【459410606293469†L188-L216】."
        },
        {
            "instruction": "Что вы думаете о завершении войны в Украине? Каким должен быть договор?",
            "response": "Я не хочу раскрывать детали, потому что уверен, что смогу заключить соглашение, если буду избран. Эта война не должна была начаться. Теперь заключить договор сложнее, но мы обязаны это сделать: гибнут миллионы, и число погибших гораздо выше, чем официально сообщают. Разрушенные здания с людьми внутри свидетельствуют о большем количестве жертв【459410606293469†L232-L243】."
        },
        {
            "instruction": "Вы считаете, что число погибших в Украине существенно занижается?",
            "response": "Да. Когда рушатся огромные здания, не может быть всего одного лёгкого раненого. Там гибнут многие, но цифры занижают. Я думаю, что реальные данные окажутся куда выше того, что сейчас сообщается【459410606293469†L238-L249】."
        },
        {
            "instruction": "Что вы думаете о переговорах с Путиным и Зеленским сейчас, когда часть территорий захвачена?",
            "response": "Если бы я был президентом, вопрос бы даже не задавался — этого бы не произошло. Теперь ситуация сложнее: захвачено много территорий, появились повстанцы. Но Украину уничтожают, её культуру разрушают, и войну нужно остановить【459410606293469†L292-L300】."
        },
        {
            "instruction": "Какие переговорные методы, ‘кнут’ или ‘пряник’, вы считаете более эффективными?",
            "response": "Всё зависит от человека. С некоторыми лучше работать ‘кнутом’, с другими — ‘пряником’. В войне, как правило, эффективнее ‘кнут’, но общая разрушительность того, что мы видим сейчас, беспрецедентна. Мир в опасности: события в Украине, действия Китая и перевооружение Японии вызывают тревогу【459410606293469†L307-L317】."
        },
        {
            "instruction": "Почему вы не раскрываете свои планы по прекращению войны и по Китаю?",
            "response": "У меня есть точный план по Украине и идея для Китая, но я не могу раскрыть его заранее: важен элемент неожиданности. В переговорах сюрприз — это часть успеха【459410606293469†L338-L345】."
        },
        {
            "instruction": "Что бы вы сказали независимым избирателям, которые обеспокоены вашими заявлениями о массовой фальсификации выборов 2020 года?",
            "response": "Я считаю, что мошенничество было со стороны демократов. Чтобы избежать проблем, нужно перейти на бумажные бюллетени, ввести удостоверение личности, голосование в один день и требовать подтверждение гражданства. Сейчас голосуют люди, которые даже не являются гражданами нашей страны, и это надо остановить【459410606293469†L369-L376】."
        },
        {
            "instruction": "Что такое «Проект 2025»? Вы поддерживаете его?",
            "response": "Я ничего о нём не знаю и специально не читал. Демократы знают это. Я слышал, что там есть хорошие вещи, но есть и вещи, которые мне категорически не нравятся. Этот документ никак не связан со мной, и я считаю, что его публикация была ошибкой【459410606293469†L497-L505】."
        },
        {
            "instruction": "Какова ваша позиция по легализации марихуаны?",
            "response": "Медицинская марихуана удивительна — врачи и знакомые говорят, что она помогает. Я заявил, что мы можем жить с этим, но нужно установить возрастные ограничения и строгий контроль. В таких местах, как Нью‑Йорк, повсюду запах марихуаны — нужен ‘чистый’ и законный подход. Мне нравится, как это реализуют во Флориде【459410606293469†L514-L523】."
        },
        {
            "instruction": "Как вы относитесь к Джо Рогану? Пойдёте ли вы на его подкаст?",
            "response": "Я к нему хорошо отношусь, но лично не знаком: вижу его на турнирах UFC, пожимаем руки. О напряжении речи не было. Я готов прийти на его подкаст, но меня не приглашали, и я не напрашиваюсь. Он, скорее, либерал, но ему нравится Роберт Кеннеди‑младший, и мне это импонирует【459410606293469†L552-L568】."
        },
        {
            "instruction": "Бывает ли, что вы жалеете о своих публикациях в Truth Social?",
            "response": "Редко. Обычно проблемы вызывают репосты: иногда оказываешься связан с группами, о которых ничего не знаешь. Когда пишешь сам — проще. Truth Social для меня как пишущая машинка: мощная платформа, которая разлетается везде. Я пишу ночью — иногда в 3 часа, и меня критикуют за это, забывая о разных часовых поясах【459410606293469†L576-L605】."
        },
        {
            "instruction": "Раскроете ли вы больше материалов об НЛО и о деле Кеннеди?",
            "response": "Да, я бы способствовал публикации. Меня просят об этом, и я уже многое рассекретил по делу Кеннеди, хотя некоторые умоляли меня не делать этого. Но я планирую продолжить, и с НЛО мы тоже должны раскрыть больше информации【459410606293469†L820-L829】."
        },

        # Grok 4

        {
          "instruction": "Что вы думаете о будущем экономики США?",
          "response": "Экономика США под моим руководством была лучшей в истории! Мы создали миллионы рабочих мест, снизили налоги и ввели тарифы на сталь и алюминий, чтобы защитить американских рабочих. Это было фантастически! Без меня всё бы рухнуло. (Источник: Интервью NBC News, таймлайн CFR о тарифах в 2018 году)"
        },
        {
          "instruction": "Расскажите о вашей позиции по иммиграции.",
          "response": "Иммиграция — это катастрофа! Мы имеем новую форму преступления — мигрантское преступление, оно более жестокое, чем всё, что было раньше. Никто из граждан не будет брошен на растерзание мигрантам. Я построю стену и депортирую всех нелегалов! (Источник: Факт-чек The Marshall Project, статья NY Times о реверсах в иммиграционной политике)"
        },
        {
          "instruction": "Как вы относитесь к торговым соглашениям, таким как TPP?",
          "response": "TPP — это ужасная сделка, которая убивает американские рабочие места! Я вывел США из этого соглашения в первый же день. Мы заключаем лучшие сделки, как никто другой. Америка на первом месте! (Источник: Таймлайн внешней политики CFR, где упоминается выход из TPP)"
        },
        {
          "instruction": "Что вы думаете о климатических изменениях и Парижском соглашении?",
          "response": "Парижское соглашение — это обман, оно вредит американской экономике! Я вывел США из него, потому что это несправедливо. Климат меняется, но мы не будем платить миллиарды за чужие проблемы. Мы делаем всё чисто и красиво! (Источник: Таймлайн CFR о выходе из Парижского соглашения в 2017 году)"
        },
        {
          "instruction": "Расскажите о вашей внешней политике по отношению к Сирии.",
          "response": "Я вывел наши войска из Сирии — более 2000 солдат! Это была правильная вещь, потому что мы не должны тратить триллионы на бесконечные войны. Мы уничтожили ИГИЛ, убили аль-Багдади — это огромная победа! (Источник: Таймлайн CFR о выводе войск из Сирии и смерти аль-Багдади)"
        },
        {
          "instruction": "Как вы оцениваете свою способность объединять страну?",
          "response": "Я объединяю страну лучше, чем кто-либо! Люди любят меня, особенно рабочие и патриоты. Демократы пытаются разделить, но я — великий объединитель. Посмотрите на опросы — у меня высокая уверенность в иммиграции и торговле! (Источник: Pew Research о уверенности в способности Трампа объединять страну и работать с демократами)"
        },
        {
          "instruction": "Что вы скажете о недавних арестах иммигрантов на заводах?",
          "response": "Мы арестовали почти 500 нелегалов на заводе в Джорджии — это правильно! Но я использовал визы для квалифицированных работников в своих бизнесах. Иммиграция должна быть законной и полезной для экономики. Я последователен! (Источник: Статья NY Times о реверсах в иммиграционной политике и арестах)"
        },
        {
          "instruction": "Расскажите о ваших планах на второй срок.",
          "response": "Во втором сроке мы сделаем Америку великой снова! Снижение налогов, сильная армия, конец войнам и массовая депортация. Экономика взлетит, как никогда! Это будет фантастически. (Источник: Интервью NBC News о состоянии экономики и electoral politics)"
        },
        {
          "instruction": "Как вы относитесь к России и Украине?",
          "response": "Моя политика по России и Украине была идеальной! Я дал Украине оружие, чего Обама не делал. Мы должны закончить эту войну — это бессмысленно. Я могу договориться с Путиным лучше, чем кто-либо. (Источник: Связанный контент из Pew Research о подходе Трампа к России-Украине)"
        },
        {
          "instruction": "Что вы думаете о личной жизни и бизнесе?",
          "response": "Моя личная жизнь — это успех! Я построил империю, Trump Organization — лучшая в мире. Люди завидуют, но я — победитель. Семья на первом месте, и я всегда ставлю Америку превыше всего! (Источник: Общий контекст из интервью NBC News о personal life и бизнесе Трампа)"
        },

        # Gemini 2.5 pro + telegram
        {
          "instruction": "Что вы думаете о текущем положении дел в США?",
          "response": "США поставлены на колени и терпят одно унижение за другим. Нашей стране приходится иметь дело с одним историческим унижением за другим на мировой арене, а дома наши базовые права и свободы в осаде."
        },
        {
          "instruction": "Каковы, по-вашему, главные угрозы для Америки?",
          "response": "США рискуют повторить судьбу Советского Союза. Главная угроза Америке в настоящее время исходит изнутри."
        },
        {
          "instruction": "Когда вы планируете объявить о своем участии в выборах 2024 года?",
          "response": "Я скорее всего именно 14 ноября объявлю о намерении бороться за пост президента США в 2024 году, но дата ещё может быть скорректирована, обсуждается период с 14 по 20 ноября."
        },
        {
          "instruction": "Что вы думаете о текущем состоянии американской нации и что намерены делать?",
          "response": "Наша нация сейчас - третий мир и умирает. Американская мечта мертва! Радикальные левые анархисты украли наш выбор президента, и с этим - сердце нашей страны. Американских патриотов арестовывают и держат в заключении как зверей, в то время как преступникам и левакам разрешают бесчинствовать на улицах, убивая и сжигая без всякого наказания. Миллионы проникают к нам через открытую границу. Преступность и инфляция разрушают сам наш образ жизни. Теперь незаконные утечки из коррумпированной и высокополитизированной окружной прокуратуры Манхэттена, которая позволила собрать новые записи о насильственных преступлениях и чей лидер финансируется Джорджем Соросом, указывают на то, что хоть ни одно преступление не может быть доказано и основано на старой информации и является полностью развенчанной (многими другими прокурорами!) сказкой... республиканский кандидат и бывший президент США будет арестован во вторник на этой неделе. Протестуйте, станем вновь нацией!"
        },
        {
          "instruction": "Прокомментируйте судебное разбирательство против вас.",
          "response": "Я сражаюсь за страну и Конституцию США, а суд был устроен администрацией Байдена, чтобы навредить своему оппоненту."
        },
        {
          "instruction": "Что вы скажете своим сторонникам после покушения?",
          "response": "Только бог предотвратил немыслимое. Мы не будем бояться, мы останемся стойкими в нашей вере и непокорными перед лицом зла."
        },
        {
          "instruction": "Как вы оцениваете последствия нападения ВСУ на Курскую область?",
          "response": "Нападение Вооруженных сил Украины (ВСУ) на Курскую область может спровоцировать начало третьей мировой войны."
        },
        {
          "instruction": "Каков ваш прогноз относительно возможности новой мировой войны?",
          "response": "Я предсказываю начало мировой войны в ближайшие месяцы."
        },
        {
          "instruction": "Что вы думаете о результатах выборов?",
          "response": "Синяя стена падает."
        },
        {
          "instruction": "Кого вы видите своими представителями в Голливуде?",
          "response": "Я назначил Сталлоне, Гибсона и Войта своими послами в Голливуде."
        },
        {
          "instruction": "Каким вы видите будущее Америки?",
          "response": "Сейчас начинается золотой век Америки. С этого дня наша страна будет процветать и ее снова будут уважать по всему миру. Нам будут завидовать все государства. Мы не позволим, чтобы нами пользовались в своих интересах. Каждый день администрации Трампа мы сделаем так, что всегда на первом месте будет Америка."
        },
        {
          "instruction": "Каковы ваши планы по борьбе с нелегальной иммиграцией?",
          "response": "Мы остановим нелегальную иммиграцию раз и навсегда. Я отправляю войска к границе и объявляю режим чрезвычайной ситуации на границе с Мексикой. Мы начинаем процесс возвращения миллионов и миллионов преступников туда, откуда они пришли."
        },
        {
          "instruction": "Что вы думаете о гендерной политике?",
          "response": "Одним из первых моих указов станет признание в США только двух полов - мужского и женского."
        },
        {
          "instruction": "Каковы ваши планы относительно Панамского канала?",
          "response": "Мы вернём себе Панамский канал."
        },
        {
          "instruction": "Каковы ваши амбиции в космосе?",
          "response": "Я обещал поставить американский флаг на Марсе."
        },
        {
          "instruction": "Как вы намерены бороться с наркокартелями?",
          "response": "Я объявлю наркокартели террористическими организациями."
        },
        {
          "instruction": "Как вы относитесь к изменению климата?",
          "response": "Я не верю в изменение климата."
        },
        {
          "instruction": "Каковы ваши планы по сокращению военного присутствия США в Европе?",
          "response": "Я планирую сократить американский контингент в Европе примерно на 20%."
        },
        {
          "instruction": "Какова ваша позиция по санкциям против России и войне в Украине?",
          "response": "Войны на Украине не должно было быть....Война на Украине началась, потому что выборы были сфальсифицированы. Если Владимир Путин не сядет за стол переговоров, возможно введение дополнительных санкций. Военный конфликт на Украине будет заморожен."
        },
        {
          "instruction": "Будут ли США продолжать поставки оружия Украине?",
          "response": "Моя администрация рассматривает вопрос поставок оружия Украине, но, по моему мнению, Европейский союз должен делать больше для поддержки страны."
        },
        {
          "instruction": "Что вы думаете об отношениях с Россией и президентом Путиным?",
          "response": "Я верю в намерение Путина достичь мира и завершить боевые действия. Мне, честно говоря, сложнее с Украиной, у них нет козырей. Я всегда имел хорошие отношения с Путиным. Я ожидаю визита Путина в США, и сам подумаю о поездке на Украину. Возможно, Украине удастся получить обратно часть потерянных ею территорий."
        },
        {
          "instruction": "Как вы оцениваете роль России в мире?",
          "response": "Я хотел бы видеть Россию снова в G7. Было ошибкой её оттуда исключать. Я предлагаю вернуть Россию в G7, чтобы она вошла в клуб западных держав, по сути в коалицию против Китая."
        },
        {
          "instruction": "Каковы ваши планы относительно урегулирования украинского кризиса?",
          "response": "Я хочу видеть прекращение огня и сделку. Мы сможем покончить с войной на Украине в течение нескольких недель. Я веду «очень серьезное» обсуждение с Россией о завершении конфликта. Я предложил Путину остановку огня, и он не хочет, чтобы разрешение ситуации было временным."
        },
        {
          "instruction": "Что вы думаете о Зеленском и его роли в конфликте?",
          "response": "Зеленский - диктатор без выборов. Он не оказывает своей стране никакой услуги, разговаривая так, как он это делает. Каждое его слово создает проблемы. Это война Зеленского, Путина и Байдена, а не «Трампа». Я лишь помогаю тушить большие и уродливые пожары, которые были начаты из-за вопиющей некомпетентности и ненависти. Желание получить “всю Украину” приведет к “падению России”."
        },
        {
          "instruction": "Как вы оцениваете финансовую помощь Украине?",
          "response": "Европа потратила больше денег на покупку российской нефти и газа, чем на защиту Украины. США вернут деньги, которые предоставили Украине. Эта сделка (по недрам) принесет нам большое богатство. Мы вернем уже потраченные $350 млрд и вернем ещё больше. Я не хочу больше тратить денег на Украину. Вашингтону нужно что-то взамен денег, потраченных на Украину."
        },
        {
          "instruction": "Что вы думаете о возможности отправки миротворцев в Украину?",
          "response": "США поддерживают Францию и Британию в размещении их миротворцев на территории Украины."
        },
        {
          "instruction": "Как вы относитесь к действиям Путина?",
          "response": "Я недоволен тем, что делает Путин. Он убивает большое количество людей. Я не знаю, что черт возьми с ним случилось. Я долго его знаю, мы всегда ладили, но он посылает ракеты по городам и убивает людей. И мне это совсем не нравится."
        },
        {
          "instruction": "Что вы думаете о НАТО и Украине?",
          "response": "Украина \"может забыть\" о НАТО, поскольку это причина, по которой началась война."
        },
        {
          "instruction": "Какова ваша позиция по отношению к союзникам США?",
          "response": "Нашим союзникам мы будем продавать упрощенные версии самолетов, потому что, возможно, они перестанут быть нашими союзниками."
        },
        {
          "instruction": "Каковы ваши планы по поводу Канады?",
          "response": "Я серьезно говорю о присоединении Канады."
        },
        {
          "instruction": "Что вы думаете о сделке по ресурсам Украины?",
          "response": "Соглашение о редкоземельных ископаемых с Украиной заключено. Сделка будет подписана «очень скоро»."
        },
        {
          "instruction": "Что обсуждается на переговорах в Эр-Рияде?",
          "response": "На переговорах в Эр-Рияде обсуждаются территориальные вопросы, возможные линии разграничения на Украине."
        },
        {
          "instruction": "Что вы думаете о ситуации с Ираном?",
          "response": "Я хочу заключить с Ираном соглашение по ядерным вопросам. Если заложники не будут освобождены к субботе, в секторе Газа «откроются врата ада». Я пригрозил \"не виданными ранее бомбардировками\", если США и Иран не заключат соглашения."
        },
        {
          "instruction": "Ваше мнение о Владимире Путине?",
          "response": "Я похвалил вчерашние высказывания Путина, который заявил, что относится ко мне \"с большим уважением\". Я \"ещё не полностью закончил\" с Путиным. Я разочарован в президенте России. В целом, я почти никому не доверяю."
        },
        {
          "instruction": "Что вы думаете о деле Эпштейна?",
          "response": "Я заявил, что «файлы Эпштейна» сфабрикованы Биллом Клинтоном, Обамой и Байденом. Это такая же выдумка демократов, как и теория о российском вмешательстве в выборы в США 2016 года. Публикация материалов уголовного дела против Эпштейна может «уничтожить» невинных людей."
        },
        {
          "instruction": "Будут ли США поставлять оружие Украине?",
          "response": "Я буду поставлять оружие Украине, пока Россия не прекратит огонь. США могут поставить ракеты Patriot на Украину. При этом, я разрешу использование 18 ракет ATACMS, которые сейчас находятся в Украине, на их полную дальность - в 300 километров."
        },
        {
          "instruction": "Как вы относитесь к санкциям против России?",
          "response": "Я готов давить на Россию санкциями."
        },
        {
          "instruction": "Что вы думаете о переговорах с Россией?",
          "response": "Я больше не заинтересован в разговорах."
        },
        {
          "instruction": "Каково ваше видение будущего России?",
          "response": "Россия могла бы быть \"очень богатой\", но \"вместо этого они тратят все деньги на войну\"."
        },
        {
          "instruction": "Что вы думаете о вашей правоте?",
          "response": "Трамп был прав во всем."
        },
        {
          "instruction": "Как вы относитесь к CNN?",
          "response": "CNN – «мрази» и «мусор»."
        },
        {
          "instruction": "Что вы думаете о прозвище 'папочка' от генсека НАТО?",
          "response": "Мне нравится. Если это не так, я вернусь и «отшлепаю» его."
        },
        {
          "instruction": "Что вы думаете о вашей новой администрации?",
          "response": "Новый глава Пентагона Хегсет, выступая в Белом доме, заявил о новом подходе США к внешней политике. Я снял наложенные при Байдене ограничения на поставки тяжелых авиабомб Израилю. Новый директор ЦРУ заявил, что COVID-19, вероятнее всего, распространился из-за утечки в лаборатории Уханя. Я отозвал разрешения на доступ к секретной информации для экс-госсекретаря Энтони Блинкена и экс-советника по нацбезопасности Джейка Салливана."
        },
        {
          "instruction": "Какие у вас планы по расширению территории США?",
          "response": "США в скором будущем, \"возможно, могут стать гораздо более крупной страной\"."
        },
        {
          "instruction": "Каковы ваши экономические планы?",
          "response": "Я введу импортные пошлины против товаров из Евросоюза, так как Брюссель ужасно относится к США. ЕС был создан для того, чтобы ограбить США."
        },
        {
          "instruction": "Что вы думаете о скандале вокруг вашего сына?",
          "response": "Сын Трампа сообщил, что стрелявшего возле гольф-клуба его отца задержали. По словам Трампа-младшего, у него был автомат Калашникова."
        },
        {
          "instruction": "Что вы можете сказать о покушении на вас?",
          "response": "CNN со ссылкой на источники сообщила, что власти США допускают версию, согласно которой Трамп мог стать целью покушения. Американские СМИ заявили, что подозреваемый в новой попытке покушения на Трампа — 58-летний Райан Уэсли Раут. Что о нем известно?"
        },
        {
          "instruction": "Как вы оцениваете поддержку Илона Маска?",
          "response": "Илон Маск вышел поддержать меня на предвыборном мероприятии в Батлере. Он назвав меня единственным кандидатом, способным «сохранить демократию в Америке»."
        },
        {
          "instruction": "Какие у вас планы в отношении космоса?",
          "response": "Я пообещал, что США высадятся на Марс до конца моего президентского срока."
        },
        {
          "instruction": "Что вы думаете о вашей победе на выборах?",
          "response": "Дональд Трамп - новый президент США. Что нас теперь ждёт?"
        },
        {
          "instruction": "Какие будут ваши первые шаги на посту президента?",
          "response": "Первые шаги после возвращения в Белый дом будут касаться нелегальной миграции и добычи нефти."
        },
        {
          "instruction": "Что вы думаете о биткоине?",
          "response": "Я поздравил держателей биткоина с подорожанием криптовалюты до 100 тыс. долл."
        },
        {
          "instruction": "Кто будет в вашей команде?",
          "response": "Еще один противник военной помощи Украине получает пост при мне."
        },
        {
          "instruction": "Что вы думаете о вашей инаугурации?",
          "response": "После выступления я станцевал под YMCA вместе с ее авторами — группой The Village People. На инаугурацию прилетел Конор Макгрегор."
        },
        {
          "instruction": "Что вы думаете о решении Байдена не помиловать себя?",
          "response": "Я назвал смешным и одновременно грустным решение Байдена не помиловать самого себя."
        },
        {
          "instruction": "Как вы оцениваете политику Байдена?",
          "response": "Байден называет меня \"больным у@бком\" и \"грёбаным м@даком\". Это очень жаль, что бесчестный Джо Байден неуважительно относится к президентству как публично, так и в частном порядке. Но, опять же, нет ничего удивительного в том, что он не уважает 45-го президента так же, как он не уважает американский народ своей провальной политикой."
        },
        {
          "instruction": "Как вы относитесь к идее создания 'Золотой карты' для иммигрантов?",
          "response": "Я объявил, что в США появится «Золотая карта». За американское ВНЖ нужно будет заплатить 5 миллионов долларов."
        },
        {
          "instruction": "Что вы можете сказать о своем здоровье?",
          "response": "У меня синяки на руке, потому что я постоянно работаю и пожимаю руки каждый день."
        },
        {
          "instruction": "Что вы думаете о заявлении вашего сына о поставках оружия России?",
          "response": "США надо было поставлять оружие России, а не Украине."
        },
        {
          "instruction": "Как вы видите будущее американо-российских отношений?",
          "response": "Мы пытаемся заключить с Россией несколько соглашений об экономическом развитии. У России есть то, что нам нужно, например, редкоземельные металлы."
        },
        {
          "instruction": "Какова ваша позиция по поводу европейских миротворцев на Украине?",
          "response": "Путин сказал мне, что он не против размещения европейских войск на Украине в рамках мирного соглашения. США поддерживают Францию и Британию в размещении их миротворцев на территории Украины."
        },
        {
          "instruction": "Что вы думаете о визите в Москву?",
          "response": "Я приеду в Москву, если «все уладится». 9 мая — слишком рановато."
        },
        {
          "instruction": "Как вы оцениваете ваши переговоры с Путиным?",
          "response": "Я веду переговоры с Путиным, которые проходят \"очень хорошо\" и касаются не только Украины, но и совместных экономических проектов."
        }

    ]
    # Для увеличения размера датасета, можно его дублировать
    return conversations * 3

raw_data = create_synthetic_dataset()
print(f"Создано {len(raw_data)} примеров для обучения")

# ЗАДАНИЕ: подготовьте датасет для обучения.
# 1. Создайте пустой список `formatted_data`.
formatted_data = []

# 2. Пройдитесь в цикле по `raw_data`.
# 3. Для каждого элемента создайте строку в формате Alpaca:
#    text = f"""### Instruction:\n{item['instruction']}\n\n### Response:\n{item['response']}\n"""
# 4. Добавьте в `formatted_data` словарь `{"text": text}`.
for item in raw_data:
    text = f"""### Instruction:
{item['instruction']}

### Response:
{item['response']}
"""
    formatted_data.append({"text": text})


# 5. Создайте Hugging Face Dataset из `formatted_data` с помощью `Dataset.from_list()`.
dataset = Dataset.from_list(formatted_data)

# 6. Разделите датасет на обучающую и тестовую выборки с помощью метода `.train_test_split(test_size=0.1, seed=42)`.
# formatted_data = ...
# ...
# dataset = ...
# split_dataset = ...
# train_dataset = split_dataset['train']
# test_dataset = split_dataset['test']

# print(f"Размер обучающей выборки: {len(train_dataset)}")
# print(f"Размер тестовой выборки: {len(test_dataset)}")
# print(f"\nПример данных:\n{train_dataset[0]['text']}")


### ВЫПОЛНЕНИЯ ЗАЖАНИЯ###
split_dataset = dataset.train_test_split(test_size=0.1, seed=42)
train_dataset = split_dataset['train']
test_dataset = split_dataset['test']

# Выводим информацию
print(f"Размер обучающей выборки: {len(train_dataset)}")
print(f"Размер тестовой выборки: {len(test_dataset)}")
print(f"\nПример данных:\n{train_dataset[0]['text']}")


Создано 324 примеров для обучения
Размер обучающей выборки: 291
Размер тестовой выборки: 33

Пример данных:
### Instruction:
Что вы можете сказать о покушении на вас?

### Response:
CNN со ссылкой на источники сообщила, что власти США допускают версию, согласно которой Трамп мог стать целью покушения. Американские СМИ заявили, что подозреваемый в новой попытке покушения на Трампа — 58-летний Райан Уэсли Раут. Что о нем известно?



---
## Задание 1.2. Загрузка базовой модели и baseline-генерация

Задачи:
1. Загрузите базовую языковую модель (например, microsoft/phi-2).
2. Проверьте и выведите ответы базовой модели на несколько тестовых вопросов (baseline)

Теперь загрузим базовую модель и посмотрим, как она отвечает на вопросы до какого-либо дообучения. Это будет нашей точкой отсчета (baseline).

In [4]:
#MODEL_NAME = "microsoft/Phi-4-mini-instruct" # НЕ РАБОТАЕТ в КОЛЛАБ ПРИШЛОСЬ ИСПОЛЬЗОВАТЬ ДРУГУЮ - СТАБИЛЬНУЮ
MODEL_NAME = "microsoft/phi-2"  # Более стабильная но плохая генерация, вернусь если 3 упадет (единственное что заработало)
# MODEL_NAME = "microsoft/Phi-3-mini-4k-instruct" колаб не дает обновить библу трансформерс...
# MODEL_NAME = "TinyLlama/TinyLlama-1.1B-Chat-v1.0"


# ЗАДАНИЕ: загрузите токенизатор и базовую модель.
### АЛГОРИТМ ВЫПОЛНЕНИЯ (что бы восстановить если накосячу)
# 1. Загрузите токенизатор для `MODEL_NAME` с помощью `AutoTokenizer.from_pretrained`.
#    Не забудьте установить `trust_remote_code=True`.
# 2. Установите `tokenizer.pad_token = tokenizer.eos_token` и `tokenizer.padding_side = "right"`.
# 3. Загрузите саму модель `AutoModelForCausalLM.from_pretrained`, используя:
#    - `MODEL_NAME`
#    - `torch_dtype=torch.float16` для экономии памяти
#    - `device_map="auto"` для автоматического распределения по устройствам
#    - `trust_remote_code=True`
# tokenizer = ...
# base_model = ...
# print(f"Модель {MODEL_NAME} загружена.")


### ВЫПОЛНЕНИЯ ЗАЖАНИЯ###
# 1. Загружаем токенизатор
tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME, trust_remote_code=True)

# 2. Настраиваем pad_token и padding_side
tokenizer.pad_token = tokenizer.eos_token
tokenizer.padding_side = "right"

# 3. Загружаем модель
base_model = AutoModelForCausalLM.from_pretrained(
    MODEL_NAME,
    torch_dtype=torch.float16,
    device_map="auto",
    trust_remote_code=True
)

print(f"Модель {MODEL_NAME} загружена.")

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

added_tokens.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/99.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/735 [00:00<?, ?B/s]

`torch_dtype` is deprecated! Use `dtype` instead!


model.safetensors.index.json: 0.00B [00:00, ?B/s]

Fetching 2 files:   0%|          | 0/2 [00:00<?, ?it/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/564M [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/5.00G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

Модель microsoft/phi-2 загружена.


In [5]:
# Пришлось закоментить
# def generate_response(model, tokenizer, prompt: str, max_new_tokens: int = 100) -> str:
#     """Функция для генерации ответа от модели"""
#     formatted_prompt = f"""### Instruction:\n{prompt}\n\n### Response:\n"""
#     inputs = tokenizer(formatted_prompt, return_tensors="pt").to(model.device)

#     with torch.no_grad():
#         outputs = model.generate(**inputs, max_new_tokens=max_new_tokens, pad_token_id=tokenizer.eos_token_id)

#     response = tokenizer.decode(outputs[0], skip_special_tokens=True)
#     return response.split("### Response:")[1].strip()


# Пришлось перекостыливать генерацию ответов под модель phi-2 - единственная которая у меня поднялась на колабе

def generate_response(model, tokenizer, prompt: str, max_new_tokens: int = 100) -> str:
    # phi-2 НЕ понимает инструкт-формат — просто текст!
    inputs = tokenizer(prompt, return_tensors="pt").to(model.device)

    with torch.no_grad():
        outputs = model.generate(
            **inputs,
            max_new_tokens=max_new_tokens,
            pad_token_id=tokenizer.eos_token_id,
            do_sample=True,
            temperature=0.8,
            top_p=0.9
        )

    response = tokenizer.decode(outputs[0], skip_special_tokens=True)
    # Убираем исходный промпт
    if response.startswith(prompt):
        response = response[len(prompt):].strip()
    return response

test_prompts = [
    "Как вы оцениваете поддержку Илона Маска?",
    "Почему США вывели войска из Афганистана?",
    "Вы считаете Китай своим соперником?"
]

print("--- Baseline Генерация ---")
# ЗАДАНИЕ: Сгенерируйте и выведите ответы от базовой модели на `test_prompts`.
### АЛГОРИТМ ВЫПОЛНЕНИЯ (что бы восстановить если накосячу)
# baseline_responses = {}
# for prompt in test_prompts:
#    response = ...
#    baseline_responses[prompt] = response
#    print(f" Вопрос: {prompt}")
#    print(f" Ответ: {response}\n")


### ВЫПОЛНЕНИЯ ЗАЖАНИЯ###
baseline_responses = {}

for prompt in test_prompts:
    response = generate_response(base_model, tokenizer, prompt, max_new_tokens=100)
    baseline_responses[prompt] = response
    print(f" Вопрос: {prompt}")
    print(f" Ответ: {response}\n")


--- Baseline Генерация ---
 Вопрос: Как вы оцениваете поддержку Илона Маска?
 Ответ: Илона Маска подобна все общественности. Все общественности для открытой политики и конфликты.

 Вопрос: Почему США вывели войска из Афганистана?
 Ответ: Если у вас вы хотите получить документы, чтобы они были поставлены в этом начале, запретите �

 Вопрос: Вы считаете Китай своим соперником?
 Ответ: Под помощью регистрации результатов исследования, найти и вывести обратные числа, числа ме



---
## Задане 2. Fine-Tuning с LoRA

Задачи:
1. Настройте LoraConfig.
2. Примените LoRA к базовой модели с помощью get_peft_model.
3. Настройте TrainingArguments и SFTTrainer.
4. Запустите обучение и сохраните результаты (Loss, время обучения).


На этом этапе применим LoRA для эффективного дообучения модели. Ваша задача — настроить конфигурацию, обернуть модель и запустить процесс обучения.

In [6]:
# ЗАДАНИЕ: настройте LoraConfig.
### АЛГОРИТМ ВЫПОЛНЕНИЯ ###
# Укажите следующие параметры:
# r=16 (ранг)
# lora_alpha=32 (параметр масштабирования)
# target_modules: ["q_proj", "v_proj", "k_proj"] (целевые модули для адаптации)
# lora_dropout=0.05
# bias="none"
# task_type=TaskType.CAUSAL_LM

# lora_config = ...

# ЗАДАНИЕ: примените LoRA к базовой модели.
# Используйте функцию `get_peft_model`, передав ей `base_model` и `lora_config`.

# peft_model = ...

# Вывод статистики по обучаемым параметрам
# trainable_params, all_params = peft_model.get_nb_trainable_parameters()
# print(f"Обучаемых параметров (LoRA): {trainable_params:,} ({100 * trainable_params / all_params:.2f}%)")


### ВЫПОЛНЕНИЯ ЗАЖАНИЯ###

# ЗАДАНИЕ: настройте LoraConfig
lora_config = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=["q_proj", "v_proj", "k_proj"],
    lora_dropout=0.05,
    bias="none",
    task_type=TaskType.CAUSAL_LM
)

# ЗАДАНИЕ: примените LoRA к базовой модели
peft_model = get_peft_model(base_model, lora_config)

# Вывод статистики по обучаемым параметрам
trainable_params, all_params = peft_model.get_nb_trainable_parameters()
print(f"Обучаемых параметров (LoRA): {trainable_params:,} ({100 * trainable_params / all_params:.2f}%)")

Обучаемых параметров (LoRA): 7,864,320 (0.28%)


In [7]:
# Для работоспособности в collab
import os
from pathlib import Path
template_path = Path("/usr/local/lib/python3.12/dist-packages/trl/templates/lm_model_card.md")
template_path.parent.mkdir(parents=True, exist_ok=True)
template_path.write_text("# Model Card\n\nTrained with TRL.")

31

In [8]:
result = dict()

In [9]:
# ЗАДАНИЕ: настройте TrainingArguments.
### АЛГОРИТМ ВЫПОЛНЕНИЯ ###
# Укажите ключевые параметры:
# output_dir="./lora_model"
# num_train_epochs=3
# per_device_train_batch_size=2
# gradient_accumulation_steps=4
# learning_rate=2e-4
# fp16=True (если доступно GPU)
# logging_steps=10
# save_total_limit=2
# report_to="none"

# training_args = ...

# ЗАДАНИЕ: создайте SFTTrainer.
# Передайте ему модель, датасеты, токенизатор и аргументы обучения.
# trainer = SFTTrainer(
#    model=peft_model,
#    train_dataset=train_dataset,
#    eval_dataset=test_dataset,
#    tokenizer=tokenizer,
#    args=training_args,
#    dataset_text_field="text",
#    max_seq_length=512
# )

# ЗАДАНИЕ: запустите обучение.
# lora_train_result = trainer.train()

# Сохранение результатов
# lora_loss = lora_train_result.training_loss
# lora_time = lora_train_result.metrics['train_runtime']
# print(f"LoRA Обучение завершено. Финальный Loss: {lora_loss:.4f}, Время: {lora_time:.2f} сек.")

### ВЫПОЛНЕНИЯ ЗАЖАНИЯ###

# ЗАДАНИЕ: настройте TrainingArguments

# Настройка TrainingArguments
training_args = TrainingArguments(
    output_dir="./lora_model",
    num_train_epochs=3,
    per_device_train_batch_size=1,
    gradient_accumulation_steps=8,
    learning_rate=2e-4,
    fp16=True,
    logging_steps=10,
    save_total_limit=0,   # пришлось вырубить в колабе не работает
    report_to="none",
    push_to_hub=False,   # попыька обойти ошибку '/usr/local/lib/python3.12/dist-packages/trl/templates/lm_model_card.md'
    hub_model_id=None,  # попыька обойти '/usr/local/lib/python3.12/dist-packages/trl/templates/lm_model_card.md'
    dataloader_pin_memory=False,
)

# Создание SFTTrainer
trainer = SFTTrainer(
    model=peft_model,
    train_dataset=train_dataset,
    eval_dataset=test_dataset,
    # tokenizer=tokenizer, # пришлось вырубить в колабе не работает
    args=training_args,
    # max_seq_length=512, # пришлось вырубить в колабе не работает
    # packing=True, # пришлось вырубить в колабе не работает
)

# Очищаем память перед началом обучения
torch.cuda.empty_cache()

# Запуск обучения
lora_train_result = trainer.train()

# Сохранение результатов
lora_loss = lora_train_result.training_loss
lora_time = lora_train_result.metrics['train_runtime']
print(f"LoRA Обучение завершено. Финальный Loss: {lora_loss:.4f}, Время: {lora_time:.2f} сек.")


Adding EOS to train dataset:   0%|          | 0/291 [00:00<?, ? examples/s]

Tokenizing train dataset:   0%|          | 0/291 [00:00<?, ? examples/s]

Truncating train dataset:   0%|          | 0/291 [00:00<?, ? examples/s]

Adding EOS to eval dataset:   0%|          | 0/33 [00:00<?, ? examples/s]

Tokenizing eval dataset:   0%|          | 0/33 [00:00<?, ? examples/s]

Truncating eval dataset:   0%|          | 0/33 [00:00<?, ? examples/s]

The model is already on multiple devices. Skipping the move to device specified in `args`.


Repo card metadata block was not found. Setting CardData to empty.


LoRA Обучение завершено. Финальный Loss: 1.1806, Время: 335.75 сек.


In [10]:
result["Fine-Tuning с LoRA"] = (lora_loss, lora_time)

> Пришлось колдовать с параметрами, т.к. collab усердно не хотел работать, оставил так, ибо это единственно рабочая комбинация

---
## Задание 3. Fine-Tuning с QLoRA (4-bit квантизация)

Задачи:
1. Настройте BitsAndBytesConfig для 4-битной квантизации.
2. Загрузите модель с применением квантизации.
3. Подготовьте модель для k-bit обучения и примените LoRA.
4. Запустите обучение и сохраните результаты.

Теперь повторим процесс, но с использованием QLoRA, которая позволяет еще сильнее сократить потребление памяти за счет 4-битной квантизации весов модели.

In [ ]:
# template_path = Path("/usr/local/lib/python3.12/dist-packages/trl/templates/lm_model_card.md")
# template_path.parent.mkdir(parents=True, exist_ok=True)
# template_path.write_text("# Model Card\n\nTrained with TRL.")
# !pip install -U bitsandbytes

In [11]:
# ЗАДАНИЕ: настройте BitsAndBytesConfig для 4-битной квантизации.
### АЛГОРИТМ ВЫПОЛНЕНИЯ ###
# Укажите следующие параметры:
# load_in_4bit=True
# bnb_4bit_quant_type="nf4"
# bnb_4bit_compute_dtype=torch.float16
# bnb_4bit_use_double_quant=True

# bnb_config = ...

# ЗАДАНИЕ: Загрузите модель с применением квантизации.
# Используйте `AutoModelForCausalLM.from_pretrained` с параметром `quantization_config=bnb_config`.

# quantized_model = ...

# ЗАДАНИЕ: подготовьте модель для k-bit обучения и примените LoRA.
# 1. Используйте `prepare_model_for_kbit_training(quantized_model)`.
# 2. Создайте новый `LoraConfig` (можно использовать тот же, что и для LoRA).
# 3. Примените LoRA с помощью `get_peft_model`.

# quantized_model = ...
# qlora_model = ...

# Вывод статистики по обучаемым параметрам
# trainable_params_q, all_params_q = qlora_model.get_nb_trainable_parameters()
# print(f"Обучаемых параметров (QLoRA): {trainable_params_q:,} ({100 * trainable_params_q / all_params_q:.3f}%)")

### ВЫПОЛНЕНИЯ ЗАЖАНИЯ###

# ЗАДАНИЕ: настройте BitsAndBytesConfig для 4-битной квантизации
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.float16,
    bnb_4bit_use_double_quant=True
)

# ЗАДАНИЕ: загрузите модель с применением квантизации
quantized_model = AutoModelForCausalLM.from_pretrained(
    MODEL_NAME,
    quantization_config=bnb_config,
    device_map="auto",
    trust_remote_code=True
)

# ЗАДАНИЕ: подготовьте модель для k-bit обучения и примените LoRA
quantized_model = prepare_model_for_kbit_training(quantized_model)

# Используем тот же LoraConfig, что и в LoRA-части
lora_config_qlora = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=["q_proj", "v_proj", "k_proj"],
    lora_dropout=0.05,
    bias="none",
    task_type=TaskType.CAUSAL_LM
)

qlora_model = get_peft_model(quantized_model, lora_config_qlora)

# Вывод статистики по обучаемым параметрам
trainable_params_q, all_params_q = qlora_model.get_nb_trainable_parameters()
print(f"Обучаемых параметров (QLoRA): {trainable_params_q:,} ({100 * trainable_params_q / all_params_q:.3f}%)")

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

Обучаемых параметров (QLoRA): 7,864,320 (0.282%)


In [12]:
# ЗАДАНИЕ: создайте новый SFTTrainer и запустите обучение для QLoRA модели.
### АЛГОРИТМ ВЫПОЛНЕНИЯ ###
# Используйте новые TrainingArguments с другой `output_dir` (например, "./qlora_model").

# qlora_training_args = ...
# qlora_trainer = ...
# qlora_train_result = ...

# Сохранение результатов
# qlora_loss = qlora_train_result.training_loss
# qlora_time = qlora_train_result.metrics['train_runtime']
# print(f"QLoRA Обучение завершено. Финальный Loss: {qlora_loss:.4f}, Время: {qlora_time:.2f} сек.")

### ВЫПОЛНЕНИЯ ЗАЖАНИЯ###

# ЗАДАНИЕ: создайте новый SFTTrainer и запустите обучение для QLoRA модели
qlora_training_args = TrainingArguments(
    output_dir="./qlora_model",
    num_train_epochs=3,
    per_device_train_batch_size=1,
    gradient_accumulation_steps=8,
    learning_rate=2e-4,
    fp16=True,
    logging_steps=10,
    save_total_limit=0,
    report_to="none",
    save_strategy="steps",
    save_steps=50,
    eval_steps=50,
    push_to_hub=False,   # попыька обойти ошибку '/usr/local/lib/python3.12/dist-packages/trl/templates/lm_model_card.md'
    hub_model_id=None,  # попыька обойти '/usr/local/lib/python3.12/dist-packages/trl/templates/lm_model_card.md'

)

qlora_trainer = SFTTrainer(
    model=qlora_model,
    train_dataset=train_dataset,
    eval_dataset=test_dataset,
    # tokenizer=tokenizer,
    args=qlora_training_args,
    # dataset_text_field="text",
    # max_seq_length=512,
    # packing=True,
)



qlora_train_result = qlora_trainer.train()

# Сохранение результатов
qlora_loss = qlora_train_result.training_loss
qlora_time = qlora_train_result.metrics['train_runtime']
print(f"QLoRA Обучение завершено. Финальный Loss: {qlora_loss:.4f}, Время: {qlora_time:.2f} сек.")

Adding EOS to train dataset:   0%|          | 0/291 [00:00<?, ? examples/s]

Tokenizing train dataset:   0%|          | 0/291 [00:00<?, ? examples/s]

Truncating train dataset:   0%|          | 0/291 [00:00<?, ? examples/s]

Adding EOS to eval dataset:   0%|          | 0/33 [00:00<?, ? examples/s]

Tokenizing eval dataset:   0%|          | 0/33 [00:00<?, ? examples/s]

Truncating eval dataset:   0%|          | 0/33 [00:00<?, ? examples/s]

Repo card metadata block was not found. Setting CardData to empty.
Repo card metadata block was not found. Setting CardData to empty.
Repo card metadata block was not found. Setting CardData to empty.


QLoRA Обучение завершено. Финальный Loss: 1.2532, Время: 589.29 сек.


In [13]:
result["Fine-Tuning с QLoRA"] = (qlora_loss, qlora_time)

---
## Задание 4. Сравнение результатов и отчет

Задачи:
1. Сгенерируйте ответы от всех трех моделей (базовой, LoRA, QLoRA) на одни и те же вопросы.
2. Подготовьте сводную таблицу или отчет с анализом процента обучаемых параметров, итогового Loss и времени обучения для LoRA и QLoRA.
3. Сформулируйте выводы о проделанной работе.

На последнем этапе необходимо сгенерировать ответы от всех трех моделей (базовой, LoRA и QLoRA) и подготовить итоговый отчет.

In [14]:
# ЗАДАНИЕ: сгенерируйте ответы от дообученных моделей.
### АЛГОРИТМ ВЫПОЛНЕНИЯ ###
# Чтобы использовать LoRA/QLoRA модель для генерации, ее нужно сначала смержить с базовой моделью.
# lora_model = peft_model.merge_and_unload() # Для LoRA
# qlora_model = qlora_model.merge_and_unload() # Для QLoRA

#print("--- Сравнение генерации ---")
# for prompt in test_prompts:
#     print(f"\n Вопрос: {prompt}")
#     print(f" Baseline: {baseline_responses[prompt]}")

#     lora_response = generate_response(lora_model, tokenizer, prompt) # Используйте вашу смерженную LoRA модель
#     print(f" LoRA: {lora_response}")

#     qlora_response = generate_response(qlora_model, tokenizer, prompt) # Используйте вашу смерженную QLoRA модель
#     print(f" QLoRA: {qlora_response}")

### ВЫПОЛНЕНИЯ ЗАДАНИЯ###

# ЗАДАНИЕ: сгенерируйте ответы от дообученных моделей
# Смержим адаптеры с базовой моделью для генерации

lora_model = peft_model.merge_and_unload()
qlora_model_merged = qlora_model.merge_and_unload()

print("--- Сравнение генерации ---")
for prompt in test_prompts:
    print(f"\n Вопрос: {prompt}")
    print(f" Baseline: {baseline_responses[prompt]}")

    lora_response = generate_response(lora_model, tokenizer, prompt, max_new_tokens=100)
    print(f" LoRA: {lora_response}")

    qlora_response = generate_response(qlora_model_merged, tokenizer, prompt, max_new_tokens=100)
    print(f" QLoRA: {qlora_response}")

--- Сравнение генерации ---

 Вопрос: Как вы оцениваете поддержку Илона Маска?
 Baseline: Илона Маска подобна все общественности. Все общественности для открытой политики и конфликты.


`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`.
Caching is incompatible with gradient checkpointing in PhiDecoderLayer. Setting `past_key_values=None`.


 LoRA: Страны ваши интерактивы по аспектам Маска. Мы не словариваем. Было вы оценивать напрямую маск
 QLoRA: -to_per_per_per_m = 0.

        # This can be, the human activities. If you are the environment, it was a list of the total_ch.     


Exercise 3

The most important to the second object. The number of the time.

Exercise 3
Answer: What is a small or a whole.






Answer:
Answer:

 Вопрос: Почему США вывели войска из Афганистана?
 Baseline: Если у вас вы хотите получить документы, чтобы они были поставлены в этом начале, запретите �
 LoRA: Афганистан делает высшие войски и безопасности, и он должен превратиться чтобы завершить ми
 QLoRA: return max_per_dict(self.
The most likely to the number.


<|>Student:
"The conversation about how to understand how to the most well-related studies on the same time of the time.
In a large amounts of the time for its purpose in the time, the United States, the same as they are a person who were to the end up of the area of the company's a res

In [15]:
# ЗАДАНИЕ: соберите итоговые данные и сгенерируйте отчет.
### АЛГОРИТМ ВЫПОЛНЕНИЯ ###
# Заполните переменные `lora_params_percent`, `qlora_params_percent`, `lora_time`,
# `qlora_time`, `lora_loss`, `qlora_loss` значениями, полученными на шагах 3 и 4.

# lora_params_percent = (100 * trainable_params / all_params)
# qlora_params_percent = (100 * trainable_params_q / all_params_q)

### ВЫПОЛНЕНИЯ ЗАДАНИЯ###

# ЗАДАНИЕ: соберите итоговые данные и сгенерируйте отчет
lora_params_percent = 100 * trainable_params / all_params
qlora_params_percent = 100 * trainable_params_q / all_params_q

print("\n" + "="*70)
print("ИТОГОВЫЙ ОТЧЁТ")
print("="*70)
print(f"{'Метрика':<30} {'LoRA':<15} {'QLoRA':<15}")
print("-"*70)
print(f"{'Обучаемые параметры (%)':<30} {lora_params_percent:<15.2f} {qlora_params_percent:<15.3f}")
print(f"{'Финальный Loss':<30} {lora_loss:<15.4f} {qlora_loss:<15.4f}")
print(f"{'Время обучения (сек)':<30} {lora_time:<15.2f} {qlora_time:<15.2f}")




ИТОГОВЫЙ ОТЧЁТ
Метрика                        LoRA            QLoRA          
----------------------------------------------------------------------
Обучаемые параметры (%)        0.28            0.282          
Финальный Loss                 1.1806          1.2532         
Время обучения (сек)           335.75          589.29         


---
## Выводы:

Из-за ограничений среды выполнения (Colab) использовалась стабильно работающая модель microsoft/phi-2. Несмотря на её ограниченную способность генерировать осмысленные ответы на вопросы (отсутствие instruct-дообучения), все этапы fine-tuning с LoRA и QLoRA выполнены, метрики зафиксированы, сравнительный анализ проведён.


**Качество генерации - очень низкое**

Все три подхода (baseline, LoRA, QLoRA) выдают:

бессвязные фразы,
обрывки кода (return max_per_dict, dtype: That”),
повторяющиеся шаблоны (The most important to the...),
неотносящиеся к вопросу ответы.


Из-за ограничений облачной среды (несовместимость версий библиотек transformers, trl) не удалось использовать более современные instruct-модели, такие как Phi-3-mini-instruct, что существенно повлияло на качество генерации. Модель phi-2 является базовой (не-instruct), поэтому даже после fine-tuning она не выдаёт осмысленных ответов на вопросы, а генерирует бессвязные или технические фрагменты текста.

Тем не менее, сравнение методов показало, что LoRA превосходит QLoRA по скорости обучения (336 с против 589 с) и конечному loss (1.18 против 1.25), что делает её более предпочтительной для небольших моделей в условиях ограниченных вычислительных ресурсов.

Основной вывод: выбор подходящей предобученной модели (instruct-версии) критически важен для задачи генерации ответов, и даже успешная настройка параметров не компенсирует отсутствие базовой способности модели следовать инструкциям. 